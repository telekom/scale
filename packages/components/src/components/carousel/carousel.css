/**
* @prop --carousel-height: Height of the carousel
* @prop --carousel-width: Width of the carousel
* @prop --carousel-slide-transition: Transition of the carousel slide
*/

.carousel {
  height: var(--carousel-height, 300px);
  width: var(--carousel-width, 400px);
  position: relative;
}

.carousel__container {
  position: relative;
  height: 100%;
  width: 100%;
  align-items: center;
  display: flex;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  overflow: hidden;
}

.carousel--vertical .carousel__container {
  flex-direction: column;
}

.carousel__slide {
  position: relative;
  height: 100%;
  min-width: 100%;
  transition: var(--carousel-slide-transition, all 0.3s ease-in-out);
}

.carousel__arrow {
  position: absolute;
  top: 50%;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  cursor: pointer;
  transform: translateY(-50%);
  border: none;
  outline: none;
  background: rgba(31, 45, 61, 0.11);
  color: #fff;
  z-index: 99;
  justify-content: center;
  display: flex;
  align-items: center;
}

.carousel--vertical .carousel__arrow {
  display: none;
}
.carousel__arrow--left {
  left: 12px;
}

.carousel__arrow--right {
  right: 12px;
}

.carousel__indicators {
  position: absolute;
  list-style: none;
  margin: 0;
  padding: 0;
  z-index: 2;
  display: inline-flex;
}

.carousel__indicator {
  display: block;
  opacity: 0.5;
  width: 30px;
  height: 4px;
  background: #fff;
  border: none;
  outline: none;
  padding: 0;
  margin: 0 6px;
  cursor: pointer;
  transition: 0.3s;
}

.carousel__indicator--active {
  opacity: 1;
}

.carousel__indicators {
  bottom: 12px;
  left: 50%;
  transform: translateX(-50%);
}

.carousel--vertical .carousel__indicators {
  left: unset;
  bottom: 0;
  right: 12px;
  top: 50%;
  transform: translateY(-50%);
  display: flex;
  flex-direction: column;
}

.carousel--vertical .carousel__indicator {
  margin: 6px 0;
  width: 4px;
  height: 30px;
}
