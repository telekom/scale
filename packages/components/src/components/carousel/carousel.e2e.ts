import { newE2EPage } from '@stencil/core/testing';

describe('t-carousel', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-carousel/>');
    const element = await page.find('t-carousel');
    expect(element).toHaveClass('hydrated');
  });
});
