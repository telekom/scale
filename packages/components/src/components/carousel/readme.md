# t-carousel



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                           | Type      | Default |
| ---------- | ---------- | ------------------------------------- | --------- | ------- |
| `vertical` | `vertical` | (optional) carousel display direction | `boolean` | `false` |


## CSS Custom Properties

| Name                          | Description                      |
| ----------------------------- | -------------------------------- |
| `--carousel-height`           | Height of the carousel           |
| `--carousel-slide-transition` | Transition of the carousel slide |
| `--carousel-width`            | Width of the carousel            |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
