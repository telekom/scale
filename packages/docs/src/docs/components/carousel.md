---
title: Carousel
description: Carousel element
url: /docs/carousel
contributors:
  - maomaoZH
  - eeegor
---

# Carousel

Carousel component

selector: `t-carousel`

## Carousel horizontal

```javascript
<t-carousel>
  <div slot="arrow-left">&larr;</div>
  <div>
    <img src="http://placehold.it/400x300/d3dce6" />
  </div>
  <div>
    <img src="http://placehold.it/400x300/99a9bf" />
  </div>
  <div>
    <img src="http://placehold.it/400x300/d3dce6" />
  </div>
  <div>
    <img src="http://placehold.it/400x300/99a9bf" />
  </div>
  <div slot="arrow-right">&rarr;</div>
</t-carousel>
```

<div class="demo-container">
  <t-carousel>
    <div slot="arrow-left">&larr;</div>
    <div><img src="http://placehold.it/400x300/d3dce6" /></div>
    <div><img src="http://placehold.it/400x300/99a9bf" /></div>
    <div><img src="http://placehold.it/400x300/d3dce6" /></div>
    <div><img src="http://placehold.it/400x300/99a9bf" /></div>
    <div slot="arrow-right">&rarr;</div>
  </t-carousel>
</div>

## Carousel vertical

```javascript
<t-carousel vertical>
  <div>
    <img src="http://placehold.it/400x300/d3dce6" />
  </div>
  <div>
    <img src="http://placehold.it/400x300/99a9bf" />
  </div>
  <div>
    <img src="http://placehold.it/400x300/d3dce6" />
  </div>
  <div>
    <img src="http://placehold.it/400x300/99a9bf" />
  </div>
</t-carousel>
```

<div class="demo-container">
  <t-carousel vertical>
    <div><img src="http://placehold.it/400x300/d3dce6" /></div>
    <div><img src="http://placehold.it/400x300/99a9bf" /></div>
    <div><img src="http://placehold.it/400x300/d3dce6" /></div>
    <div><img src="http://placehold.it/400x300/99a9bf" /></div>
  </t-carousel>
</div>
