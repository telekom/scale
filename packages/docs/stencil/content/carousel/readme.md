# scale-carousel



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                           | Type                                     | Default     |
| ------------- | -------------- | ------------------------------------- | ---------------------------------------- | ----------- |
| `customClass` | `custom-class` | (optional) Carousel class             | `string`                                 | `''`        |
| `styles`      | --             | (optional) Injected jss styles        | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `vertical`    | `vertical`     | (optional) carousel display direction | `boolean`                                | `false`     |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
