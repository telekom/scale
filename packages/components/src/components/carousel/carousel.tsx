import { Component, Prop, h, State, Element, Host } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { CssInJs } from '../../utils/css-in-js';
import { styles } from './carousel.styles';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

@Component({
  tag: 'scale-carousel',
  shadow: true,
})
export class Carousel implements Base {
  @Element() hostElement: HTMLElement;
  /** (optional) Carousel class */
  @Prop() customClass?: string = '';
  /** (optional) carousel display direction */
  @Prop() vertical?: boolean = false;

  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Carousel', styles) stylesheet: StyleSheet;

  @State() slidesArray = [];
  @State() value = 0;

  componentWillLoad() {
    if (this.slidesArray.length === 0) {
      const children = this.hostElement.children;
      // tslint:disable-next-line: prefer-for-of
      for (let childIndex = 0; childIndex < children.length; childIndex++) {
        if (children[childIndex].slot === '') {
          // tslint:disable-next-line: prefer-for-of
          for (
            let slideIndex = 0;
            slideIndex < children[childIndex].children.length;
            slideIndex++
          ) {
            const element = children[childIndex].children[slideIndex];
            this.slidesArray.push(element);
          }
        }
      }
    }
  }
  componentWillUpdate() {}
  componentDidUnload() {}

  handleSlideChange = direction => {
    const val = this.value;
    if (direction === 'prev') {
      val === 0
        ? (this.value = -100 * (this.slidesArray.length - 1))
        : (this.value = val + 100);
    }

    if (direction === 'next') {
      val === -100 * (this.slidesArray.length - 1)
        ? (this.value = 0)
        : (this.value = val - 100);
    }
  };

  setActiveSlide = index => {
    this.value = -100 * index;
  };

  setTransformValue = () => {
    if (!!this.vertical) {
      return `translateY(${this.value}%)`;
    }
    return `translateX(${this.value}%)`;
  };

  setActiveCssClass = index => {
    if (Math.abs(this.value) / 100 === index) {
      return 'carousel__indicator--active';
    }
    return '';
  };

  render() {
    const { classes } = this.stylesheet;
    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <div class={this.getCssClassMap()}>
          <div class={classes.carousel__container}>
            <div
              class={`${classes.carousel__arrow} ${classes['carousel__arrow--left']}`}
              onClick={() => this.handleSlideChange('prev')}
            >
              <slot name="arrow-left" />
            </div>
            {this.slidesArray.map(element => (
              <div
                class={classes.carousel__slide}
                style={{ transform: this.setTransformValue() }}
              >
                <div innerHTML={element.outerHTML}></div>
              </div>
            ))}
            <div
              class={`${classes.carousel__arrow} ${classes['carousel__arrow--right']}`}
              onClick={() => this.handleSlideChange('next')}
            >
              <slot name="arrow-right" />
            </div>
          </div>
          <ul class={classes.carousel__indicators}>
            {Array.from(Array(this.slidesArray.length).keys()).map(index => (
              <li
                key={index}
                class={`${classes.carousel__indicator} ${this.setActiveCssClass(
                  index
                )}`}
                onClick={() => this.setActiveSlide(index)}
              ></li>
            ))}
          </ul>
        </div>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.carousel,
      this.customClass && this.customClass,
      this.vertical && classes['carousel--vertical']
    );
  }
}
