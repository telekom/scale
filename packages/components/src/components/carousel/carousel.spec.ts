import { newSpecPage } from '@stencil/core/testing';
import { Carousel } from './carousel';

describe('Carousel', () => {
  let element;
  beforeEach(async () => {
    element = new Carousel();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Carousel],
      html: `<t-carousel>Label</t-carousel>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should set correct translate value when clicking arrow prev', () => {
    element.value = 0;
    element.slidesArray = [0, 1, 2, 3];
    element.handleSlideChange('prev');
    expect(element.value).toBe(-300);
  });

  it('should set correct translate value when clicking arrow next', () => {
    element.value = 100;
    element.slidesArray = [0, 1, 2, 3];
    element.handleSlideChange('next');
    expect(element.value).toBe(0);
  });

  it('should set correct active slide', () => {
    element.value = -300;
    element.setActiveSlide(2);
    expect(element.value).toBe(-200);
  });

  it('should set correct transform value based on the carousel container direction', () => {
    element.value = -300;
    element.vertical = true;
    element.setTransformValue();
    expect(element.setTransformValue()).toBe('translateY(-300%)');
  });

  it('should set correct transform value based on the carousel container direction', () => {
    element.value = -200;
    element.setTransformValue();
    expect(element.setTransformValue()).toBe('translateX(-200%)');
  });

  it('should set css active class', () => {
    element.value = -200;
    element.setActiveCssClass(2);
    expect(element.setActiveCssClass(2)).toBe('carousel__indicator--active');
  });
});
