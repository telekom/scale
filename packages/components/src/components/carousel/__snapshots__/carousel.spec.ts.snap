// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should match snapshot 1`] = `
<scale-carousel>
  <mock:shadow-root>
    <style>
      .carousel-0-0-11 {
  width: 400px;
  height: 300px;
  position: relative;
}
.carousel__container-0-0-12 {
  width: 100%;
  height: 100%;
  margin: 0;
  display: flex;
  padding: 0;
  overflow: hidden;
  position: relative;
  box-sizing: border-box;
  align-items: center;
}
.carousel__slide-0-0-13 {
  height: 100%;
  position: relative;
  min-width: 100%;
  transition: all 0.3s ease-in-out;
}
.carousel__arrow-0-0-14 {
  top: 50%;
  color: #fff;
  width: 24px;
  border: none;
  cursor: pointer;
  height: 24px;
  display: flex;
  outline: none;
  z-index: 99;
  position: absolute;
  transform: translateY(-50%);
  background: rgba(31, 45, 61, 0.11);
  align-items: center;
  border-radius: 50%;
  justify-content: center;
}
.carousel__arrow--left-0-0-15 {
  left: 12px;
}
.carousel__arrow--right-0-0-16 {
  right: 12px;
}
.carousel__indicators-0-0-17 {
  left: 50%;
  bottom: 12px;
  margin: 0;
  display: inline-flex;
  padding: 0;
  z-index: 2;
  position: absolute;
  transform: translateX(-50%);
  list-style: none;
}
.carousel__indicator-0-0-18 {
  width: 30px;
  border: none;
  cursor: pointer;
  height: 4px;
  margin: 0 6px;
  display: block;
  opacity: 0.5;
  outline: none;
  padding: 0;
  background: #fff;
  transition: 0.3s;
}
.carousel__indicator--active-0-0-19 {
  opacity: 1;
}
.carousel--vertical-0-0-20 {}
.carousel--vertical-0-0-20 .carousel__container-0-0-12 {
  flex-direction: column;
}
.carousel--vertical-0-0-20 .carousel__arrow-0-0-14 {
  display: none;
}
.carousel--vertical-0-0-20 .carousel__indicators-0-0-17 {
  top: 50%;
  left: unset;
  right: 12px;
  bottom: 0;
  display: flex;
  transform: translateY(-50%);
  flex-direction: column;
}
.carousel--vertical-0-0-20 .carousel__indicator-0-0-18 {
  width: 4px;
  height: 30px;
  margin: 6px 0;
}
    </style>
    <div class="carousel-0-0-11">
      <div class="carousel__container-0-0-12">
        <div class="carousel__arrow--left-0-0-15 carousel__arrow-0-0-14">
          <slot name="arrow-left"></slot>
        </div>
        <div class="carousel__arrow--right-0-0-16 carousel__arrow-0-0-14">
          <slot name="arrow-right"></slot>
        </div>
      </div>
      <ul class="carousel__indicators-0-0-17"></ul>
    </div>
  </mock:shadow-root>
  Label
</scale-carousel>
`;
